### PR TITLE
[SYCL][Level-Zero] Fixes to ZE_CALL tracing and introduction of internal PI_CALL tracing.

### PR DESCRIPTION
**Pull request by smaslov-intel**
It also improves error handling by not silently ignoring errors occurring in internally made PI calls.
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -73,6 +73,53 @@ std::mutex ZeCall::GlobalLock;
...
+  }
+  return It->second;
+}
+
+// Trace a call to Level-Zero RT
+#define ZE_CALL(Call)                                                          \
+  if (auto Result = ZeCall().doCall(Call, #Call, true))                        \
+    return mapError(Result);
+#define ZE_CALL_NOCHECK(Call) ZeCall().doCall(Call, #Call, false)
+
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:115]: This is just moved intact from its current location.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -3635,7 +3640,8 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
 pi_result piEventCreate(pi_context Context, pi_event *RetEvent) {
   size_t Index = 0;
   ze_event_pool_handle_t ZeEventPool = {};
-  ZE_CALL(Context->getFreeSlotInExistingOrNewPool(ZeEventPool, Index));
+  Context->getFreeSlotInExistingOrNewPool(ZeEventPool, Index);
```
Comment (v-klochkov) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3643]: The ze_result_t return value of this method is always ignored.
Shouldn't the method return void then? Or changing the return from ze_result_t to void breaks backward compatibility?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -3892,13 +3899,13 @@ pi_result piEventRelease(pi_event Event) {
     ZE_CALL(zeEventDestroy(Event->ZeEvent));
 
     auto Context = Event->Context;
-    ZE_CALL(Context->decrementAliveEventsInPool(Event->ZeEventPool));
+    Context->decrementAliveEventsInPool(Event->ZeEventPool);
```
Comment (v-klochkov) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3902]: Same question as for Context->getFreeSlotInExistingOrNewPool() above.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Added error check

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: Added error check

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -3635,7 +3640,9 @@ piEnqueueKernelLaunch(pi_queue Queue, pi_kernel Kernel, pi_uint32 WorkDim,
 pi_result piEventCreate(pi_context Context, pi_event *RetEvent) {
   size_t Index = 0;
   ze_event_pool_handle_t ZeEventPool = {};
-  ZE_CALL(Context->getFreeSlotInExistingOrNewPool(ZeEventPool, Index));
+  if (auto Res = Context->getFreeSlotInExistingOrNewPool(ZeEventPool, Index))
```
Comment (v-klochkov) [inline sycl/plugins/level_zero/pi_level_zero.cpp:3643]: getFreeSlotInExistingOrNewPool() returns ze_result_t, while the result of piEventCreate is pi_result. It looks like that 'mapError' map should be applied to Res, right?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:3643]: No. I've changed the 2 methods to return pi_result instead.

**Review by v-klochkov — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 5]
```diff
@@ -239,12 +239,11 @@ struct _pi_context : _pi_object {
 
   // Get index of the free slot in the available pool. If there is no avialble
```
Comment (bader) [inline sycl/plugins/level_zero/pi_level_zero.hpp:240]: Please, remove executable flag from the header file.

**Review by v-klochkov — APPROVED** [review]

**Pull request merged**